File: src/lib/providers/seedream.js
import { seedreamSize } from "../seedream";
import { providerModelId } from "../model-registry";
export function seedreamPayload({ prompt, resolution, aspectRatio, references = [] }, env = process.env) {
  if (!prompt?.trim()) throw new Error("Prompt is required.");
  if (references.length > 10) throw new Error("Seedream accepts at most 10 references.");
  return { model: providerModelId("seedream-5-pro", env), prompt, size: seedreamSize(resolution, aspectRatio), response_format: "url", output_format: "png", watermark: false, optimize_prompt_options: { mode: "standard" }, ...(references.length ? { image: references } : {}) };
}
export async function generateImageSeedream(input, { signal, fetchImpl = fetch, env = process.env } = {}) {
  if (!env.ARK_API_KEY || env.ARK_API_KEY === "[SENSITIVE]") throw new Error("Seedream is not configured: ARK_API_KEY is missing or redacted.");
  const body = seedreamPayload(input, env);
  // Exactly one submission. Ambiguous transport errors/timeouts must not retry.
  let response;
  try {
    response = await fetchImpl(`${(env.ARK_BASE_URL || "https://ark.ap-southeast.bytepluses.com/api/v3").replace(/\/$/, "")}/images/generations`, {
      method: "POST", headers: { Authorization: `Bearer ${env.ARK_API_KEY}`, "Content-Type": "application/json" }, body: JSON.stringify(body), signal,
    });
  } catch { throw new Error("Seedream connection ended without a confirmed result. The request may have been charged; it was not resubmitted."); }
  const result = await response.json().catch(() => null);
  if (!response.ok || result?.error) {
    const code = String(result?.error?.code ?? "provider_error").replace(/[^a-zA-Z0-9_.-]/g, "").slice(0, 100);
    const message = code === "ModelNotOpen" ? "Enable Seedream 5.0 Pro in this account’s ModelArk console." : response.status === 429 ? "Seedream rate limit reached. Try again later." : response.status === 401 || response.status === 403 ? "Seedream authentication or model access failed." : "Seedream rejected the request. Check reference files and content requirements.";
    throw Object.assign(new Error(`${message} (${response.status}, ${code})`), { code, status: response.status });
  }
  const url = result?.data?.[0]?.url;
  if (result?.data?.length !== 1 || typeof url !== "string" || !url.startsWith("https://")) throw new Error("Seedream returned no valid single-image result. The request was not resubmitted.");
  const downloaded = await fetchImpl(url, { signal });
  if (!downloaded.ok) throw new Error(`Seedream produced an image but download failed (${downloaded.status}). The request was not resubmitted.`);
  return Buffer.from(await downloaded.arrayBuffer());
}

File: backend/apps/generation/providers/seedream.py
import os
import time
import requests
from ..seedream import size

def payload(prompt, resolution=None, aspect_ratio=None, references=None):
    references = references or []
    if not prompt.strip() or len(references) > 10:
        raise ValueError("Seedream requires a prompt and at most 10 references.")
    body = {"model": os.environ.get("SEEDREAM_MODEL") or "dola-seedream-5-0-pro-260628", "prompt": prompt, "size": size(resolution, aspect_ratio), "response_format": "url", "output_format": "png", "watermark": False, "optimize_prompt_options": {"mode": "standard"}}
    if references:
        body["image"] = references
    return body

def generate(prompt, resolution=None, aspect_ratio=None, references=None):
    key = os.environ.get("ARK_API_KEY")
    if not key or key == "[SENSITIVE]":
        raise ValueError("Seedream is not configured: ARK_API_KEY is missing or redacted.")
    body = payload(prompt, resolution, aspect_ratio, references)
    started = time.monotonic()
    try:
        response = requests.post((os.environ.get("ARK_BASE_URL") or "https://ark.ap-southeast.bytepluses.com/api/v3").rstrip("/") + "/images/generations", headers={"Authorization": f"Bearer {key}"}, json=body, timeout=(10, 210))
    except requests.RequestException as exc:
        raise RuntimeError("Seedream connection ended without a confirmed result. It may have been charged; it was not resubmitted.") from exc
    try:
        result = response.json()
    except ValueError as exc:
        raise RuntimeError("Seedream returned malformed JSON. The request was not resubmitted.") from exc
    if not response.ok or result.get("error"):
        raise RuntimeError(f"Seedream rejected the request (HTTP {response.status_code}). Check model access, rate limits, references and content requirements.")
    images = result.get("data") or []
    if len(images) != 1 or not isinstance(images[0].get("url"), str) or not images[0]["url"].startswith("https://"):
        raise RuntimeError("Seedream returned no valid single-image result. The request was not resubmitted.")
    remaining = 260 - (time.monotonic() - started)
    if remaining <= 0:
        raise RuntimeError("Seedream exceeded the queue deadline. The request was not resubmitted.")
    try:
        fetched = requests.get(images[0]["url"], timeout=(10, min(45, remaining)))
    except requests.RequestException as exc:
        raise RuntimeError("Seedream image download failed. The request was not resubmitted.") from exc
    if not fetched.ok:
        raise RuntimeError("Seedream produced an image but download failed. The request was not resubmitted.")
    return fetched.content
